### PR TITLE
feat(datasets): Add security review skill for dataset code scanning

### PR DESCRIPTION
**Pull request by ElenaKhaustova**
## Description

Implemented `kedro-plugins-security-review`, a Cursor/Copilot agent security skill that scans dataset code for unsafe deserialization, undocumented risk, and unjustified security suppressions. The skill supports two modes: full codebase scans for periodic review and PR scans for reviewing new dataset contributions before merge.

### Skill structure

```
.agents/
├── scripts/
│   └── post_github_review.sh          # Shared helper for posting PR reviews via gh API
└── skills/
    └── kedro-plugins-security-review/
        ├── SKILL.md                    # 8-step workflow, scan modes, report format
        ├── reference.md               # Dataset trust model, triage buckets, manual checks
        └── rules/
            └── kedro-plugins-security-patterns.yml   # Custom Semgrep rules
```

### Custom Semgrep rules (10 total)

7 dataset-specific rules:
- `datasets-torch-load-no-weights-only` (ERROR) — `torch.load()` without `weights_only=True`
- `datasets-tf-load-model-no-safe-mode` (ERROR) — `tf.keras.models.load_model()` without `safe_mode=True`
- `datasets-joblib-load` (WARNING) — any `joblib.load()` call
- `datasets-dynamic-read-dispatch` (WARNING) — dynamic `getattr(mod, f"read_/scan_/to_{fmt}")` that can reach `read_pickle`
- `datasets-hub-remote-code` (WARNING) — `load_dataset()` without `trust_remote_code=False`
- `datasets-transformers-pipeline` (WARNING) — `pipeline()` without `trust_remote_code=False`
- `datasets-scipy-loadmat` (WARNING) — `scipy.io.loadmat()` can deserialise pickled objects in MAT v5 files

3 rules carried from the kedro security skill:
- `kedro-yaml-unsafe-load` (WARNING) — `yaml.load()` without safe Loader
- `kedro-subprocess-shell-injection` (ERROR) — `subprocess` with `shell=True` and non-literal command
- `kedro-dynamic-import` (WARNING) — `importlib.import_module()` with non-literal argument

### Dataset trust model

The skill triages findings into 3 classification buckets:
- `dataset_vulnerability` — dataset code is unsafe, misleading, or silently drops safety args
- `by_design_with_documentation` — risk is inherent or user-configured and documented
- `needs_manual_review` — needs human judgment

### Manual review checks (run after Semgrep, even with zero findings)

1. **load_args / save_args passthrough** — are they actually reaching the underlying library call?
2. **Docstring accuracy** — do safety claims match the implementation?
3. **Risk documentation** — do binary deserializers warn about untrusted files?
4. **Security suppression audit** — are `# nosec` / `# nosemgrep` comments justified and documented?

### Scan scope

Defaults to `kedro-datasets/` (including `kedro_datasets_experimental/`). Other plugins can be scanned if explicitly requested but the triage guidance is designed for dataset code.

## Full codebase scan results

Ran the skill against `kedro-datasets/`. 25 Semgrep findings + 4 manual checks produced the following:

### Classification summary
- **dataset_vulnerability:** 4
- **needs_manual_review:** 5
- **by_design_with_documentation:** 16

### Issues found

**1. MLRun datasets silently drop `load_args` (dataset_vulnerability / ERROR)**

`MLRunModel` accepts `load_args` but `load()` calls `joblib.load(local_path)` without passing them. `MLRunResult` also accepts but ignores `load_args`. Additionally, `MLRunModel` has no docstring warning about untrusted files.

→ Follow-up: https://github.com/kedro-org/kedro-plugins/pull/1447

**2. Undocumented `# nosec` suppressions (needs_manual_review)**

5 of 6 production `# nosec` comments lack an adjacent justification:
- `pandas/gbq_dataset.py:157` — no rule ID, no explanation
- `huggingface/hugging_face_dataset.py:52` — TODO referencing #1131 but no safety rationale
- `huggingface/_base.py:157` — no explanation
- `databricks/_base_table_dataset.py:565` — `# nosec B608` with no explanation
- `polars/polars_database_dataset.py:332` — `# nosec B608` with no explanation

→ Follow-up: https://github.com/kedro-org/kedro-plugins/pull/1450

**3. TensorFlowModelDataset missing `safe_mode=True` default (dataset_vulnerability / ERROR)**

`tf.keras.models.load_model()` is called without `safe_mode=True` at lines 149 and 151. Without this flag, Keras deserialises custom objects which can execute arbitrary code. Same pattern as PyTorch `weights_only=True`. Available since TensorFlow 2.12.

→ Follow-up: https://github.com/kedro-org/kedro-plugins/pull/1448

**4. Missing risk documentation on binary deserializer datasets (dataset_vulnerability / WARNING)**

7 datasets use deserialization methods capable of arbitrary code execution but their docstrings do not mention this:
- `PickleDataset` (file and Redis)
- `TensorFlowModelDataset`
- `DartsTorchModelDataset`
- `MatlabDataset` (MAT v5 files can contain pickled objects)
- `GenericDataset` (pandas and geopandas, when `file_format: pickle`)

→ Follow-up: https://github.com/kedro-org/kedro-plugins/pull/1451

### Not actionable (by_design_with_documentation)

- `datasets-torch-load-no-weights-only` on `PyTorchDataset` — already fixed, `weights_only=True` default is correctly passed through
- `datasets-dynamic-read-dispatch` (8 hits) — pandas/geopandas/polars/ibis generic datasets; dynamic dispatch is their documented purpose
- `datasets-hub-remote-code` (2 hits) — HuggingFace datasets pass through user kwargs
- `datasets-transformers-pipeline` (1 hit) — `transformer_pipeline_dataset.py` passes through user kwargs
- `datasets-scipy-loadmat` (1 hit) — `MatlabDataset` uses `loadmat`; risk documented in follow-up ticket
- `kedro-dynamic-import` (9 hits) — pickle/redis/safetensors backends; user-controlled catalog config by design
- `pytorch_dataset.py` `# nosec: B614` — properly justified with adjacent comment

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Updated `jsonschema/kedro-catalog-X.XX.json` if necessary
- [ ] Added a description of this change in the relevant `RELEASE.md` file
- [ ] Added tests to cover my changes
- [ ] Received approvals from at least half of the TSC (required for adding a new, non-experimental dataset)

### CODE REVIEW

**Review by ankatiyar — APPROVED** [review]

**Review by merelcht — APPROVED** [review]
LGTM! 👍 

One idea I had was to add a short readme that describes how someone should use the skill. e.g. what prompts to use and at what stage of the process.

**Comment by ElenaKhaustova** [comment]
> LGTM! 👍
> 
> One idea I had was to add a short readme that describes how someone should use the skill. e.g. what prompts to use and at what stage of the process.

Sounds good, I will add a readme for both skills we have now in a separate PR.
